### PR TITLE
exp: openai integration

### PR DESCRIPTION
**Pull request by acbart**
This seems like the start of the right track, taking advantage of the Pedal API the right way. I have some small suggestions I'll place inline!

### CODE REVIEW

File: pedal/environments/blockpy.py [hunk 1]
```diff
@@ -79,6 +80,8 @@ def __init__(self, files=None, main_file='answer.py', main_code=None,
                 start_trace()
             student = run(report=report, threaded=threaded)
             student.threaded = threaded
+        if not skip_gpt:
+            gpt_run_prompts(gpt_get_api_key(openai_api_key), report=report)
```
Comment (acbart) [inline pedal/environments/blockpy.py]: I'm not entirely convinced that we want to run the prompt as a default pipeline of all the environments. In fact, I'm fairly convince that we do NOT want that at all. The only reason to add something to the default pipeline is if you are 99% certain that the majority of problems will want the default configuration for that tool. It's almost impossible to imagine a situation where the instructor does not want to verify the syntactical correctness of the student's code, so the Source tool's `verify` function is in every pipeline. 90-99% of the time, we also want to run the student's code and report runtime errors, so we call the Sandbox tool's `run` function in every pipeline (but with a boolean flag to allow special circumstances like projects to skip that step). The most borderline tool so far has been TIFA, which I'm stunned works as well as it does but is likely to be replaced with Mypy someday.

But for tools that you want heavy control over, like the Assertions, you don't see them as part of the default pipeline. The instructor calls them in the instructor control script AFTER the rest of the default pipeline executes. I think that GPT will work the same way. We're going to want to tune the prompt in some circumstances, or at least make it easy to play with the prompt. Having it in the pipeline instead of the instructor control script means you have absolute zero way to easily play with it. I know we'll stick to a lot of defaults, but that's not incompatible with having it in the ICS instead of the pipeline.

It might also be relevant for the A/B testing, where we'll probably want more direct access to the FF from the ICS. I'm still working on this feature, but I plan to make it so that you can make feedback (deterministically) randomly applied. This is key for the A/B testing we want to do. Using the students' ID as the deciding factor and some weights, you'll be able to say "I want X% of my class to get this feedback, but not the rest". See more here: https://github.com/pedal-edu/pedal/issues/98

So basically, I propose getting rid of the `skip_gpt` stuff and not running it here in the default environment. The `openai_api_key` should probably actually be some kind of `set_openai_api_key` that updates the Tool data. Then, in any ICS script you have, you would manually call `gpt_run_prompts()` (possibly with configuration parameters).

File: pedal/gpt/__init__.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+from pedal.gpt.setup import reset
+from pedal.gpt.commands import gpt_get_api_key, gpt_run_prompts
+
+NAME = 'GPT'
+SHORT_DESCRIPTION = "Runs code and other tool results through an OpenAI LLM"
+DESCRIPTION = '''
+'''
+REQUIRES = []
+OPTIONALS = []
+CATEGORY = "highest"
```
Comment (acbart) [inline pedal/gpt/__init__.py]: "highest" is not a category, it's a priority. I also think we probably want to go with either the "Instructor" or "Mistakes" category? Or propose a new category if it's really something above Syntax. What *is* the condition being detected here? I think it might actually be something similar to Mistakes, the more I think about it - that category was Luke's source code pattern matching. This is a more evolved form of that that is (theoretically) robust against syntax errors (which his was not - you can't do source code pattern matching without a parsable AST). So maybe there's a new category of something like... "LLM"? I want a better name, probably something from the NLP word that captures the fact that the condition is kind of just a dumb bag of tokens that it is treating like a black box... But anyway, I think we create that category, put it above Syntax in the priority list, and use that. Then instructors can also decide to lower the priority manually for certain prompts, or with other sophisticated schemes.

File: pedal/gpt/commands.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+try:
+    import openai
+except ImportError:
+    print("OpenAI library not found!")
+
+__all__ = ['gpt_get_api_key', 'gpt_run_prompts']
+
+
+def gpt_get_api_key(default):
```
Comment (acbart) [inline pedal/gpt/commands.py]: Either replacing or in addition to this, we probably want a function that can set the API key to be part of the Tool's data in the report.

File: pedal/gpt/commands.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            attach results to.
+    """
+    if not code:
+        code = report.submission.main_code
+
+    prompt = f"""{code}"""
+
+    messages = [
+        {'role': 'system',
+         'content': "You are an intelligent tutor for a introductory computer science course in Python. You never give answers but do give helpful tips guide students with their code."},
```
Comment (acbart) [inline pedal/gpt/commands.py]: This should probably be dragged out into a module constant. That way instructors have more control over changing these default values, and the function is a little less massive. Obviously, we'll want to eventually provide parameters for messing with the prompt in various ways, and possibly even some more functions (that change parameters stored in the Tool data for the report). But for now, I think at least moving it out of the function will be some cleaner code.

File: pedal/gpt/commands.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        {'role': 'user', 'content': prompt}
+    ]
+
+    #print(prompt)
+    response = openai.ChatCompletion.create(
+        model='gpt-3.5-turbo-0613',  # must add -0613 for functions??
```
Comment (acbart) [inline pedal/gpt/commands.py]: Should be a constant configurable in the tool's report data.

File: pedal/gpt/commands.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        top_p=0.5
+    )
+
+    #print(second_response)
+    second_response_message = second_response.choices[0].message.function_call.arguments
+
+    if second_response.choices[0].message.function_call:
+        # todo: handle incorrect responses
+        if second_response.choices[0].message.function_call.name == 'gpt_prompt_feedback':
+            args = json.loads(second_response_message)
```
Comment (acbart) [inline pedal/gpt/commands.py]: Definitely want to try/catch this with a JSONDecodeError and handle it in some way. Working on the logging stuff now, so that will be timely. But I suspect that if GPT fails (and if all retries fail, assuming we go down that route), then we want to avoid making any feedback.

File: pedal/gpt/feedbacks.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    """
+
+    Args:
+        instructions:
+    """
+
+    category = FeedbackCategory.INSTRUCTOR
+    kind = FeedbackKind.HINT
+    justification = 'Prompt Feedback'
+    title = 'Prompt Feedback'
```
Comment (acbart) [inline pedal/gpt/feedbacks.py]: Is this what you want the students to see? I would think something more like, "Intelligent Tutor Feedback" or something? This is a strongly a UX question, so I ultimately leave it to the research team and not to me :)

File: pedal/gpt/feedbacks.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    priority = "highest"
+    tool = TOOL_NAME
+    valence = 0
```
Comment (acbart) [inline pedal/gpt/feedbacks.py]: The default valence is almost certainly going to be negative here. Especially if you are doing things with the score. But I could imagine asking GPT for its valence. Basically, asking it "Is your feedback inherently positive or negative?"

File: pedal/gpt/feedbacks.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    score = 0
```
Comment (acbart) [inline pedal/gpt/feedbacks.py]: Probably better to leave the `score` as `None` by default (aka just delete this line).

File: pedal/gpt/setup.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    Remove all settings for the OpenAI tool.
+
+    Args:
+        report: The report object
+
+    Returns:
+        This tool's data
+    """
+    report[TOOL_NAME] = {
+        'feedback': []
```
Comment (acbart) [inline pedal/gpt/setup.py]: You don't need this. The feedback is a special area of the report that already exists. Instead, I believe the tool data that I've proposed is:

* `openai_key`
* Various prompt settings, or perhaps just the prompt exactly
* Prompt model, temperature, any other settings that might be interesting

In general, the tool data is "stuff we need to remember between execution of FFs." The sandbox needs to remember the results of previous executions, CAIT uses it to cache the parsed code, etc.

File: pedal/gpt/__init__.py (hunk 2 above)
Comment (acbart) [reply pedal/gpt/__init__.py]: @JAromando particularly interested if your background in the LLM stuff gives you any ideas here for a name here. Let me know if you need more explanation of what I'm looking for, it's a little bit of a philosophical conversation but also kind of important for us to be able to identify "how we know there was an issue in student's code". It's not looking at a parse tree, or execution results, or something like that... How does it "know" that there's an issue?

**Review by acbart — COMMENTED** [review]
Here's my comments!

File: pedal/gpt/__init__.py (hunk 2 above)
Comment (craftablescience) [reply pedal/gpt/__init__.py]: Ideas proposed:

- [ ] BAG_OF_WORDS
- [x] PATTERNS

File: pedal/gpt/commands.py (hunk 4 above)
Comment (craftablescience) [reply pedal/gpt/commands.py]: The reason this is inside the function is because it relies on the student code passed to the function. We can discuss tomorrow how to move it out if feasible, ~~I can't think of an easy and good way to do that~~ I wasn't thinking with functions

File: pedal/gpt/feedbacks.py (hunk 7 above)
Comment (craftablescience) [reply pedal/gpt/feedbacks.py]: Renamed to "AI-Generated Feedback", very open to discussion on this

File: pedal/gpt/setup.py (hunk 10 above)
Comment (craftablescience) [reply pedal/gpt/setup.py]: The prompt is kind of hardcoded right now so it can choose whether or not to add feedback to the report, we could have a lambda function per prompt to process the results if it's successful

**Comment by craftablescience** [comment]
Using this code to run the gpt tool now from the instructor script:
```py
from pedal import *

verify()

from pedal.gpt import gpt_run_prompts
gpt_run_prompts(temp_debug_remove_me=True)

from pedal.resolvers import print_resolve
print_resolve()
```

`temp_debug_remove_me` is a temporary argument to print debug info

**Comment by craftablescience** [comment]
> Using this code to run the gpt tool now from the instructor script:
> 
> ```python
> from pedal import *
> 
> verify()
> 
> from pedal.gpt import gpt_run_prompts
> gpt_run_prompts(temp_debug_remove_me=True)
> 
> from pedal.resolvers import print_resolve
> print_resolve()
> ```
> 
> `temp_debug_remove_me` is a temporary argument to print debug info

Removed that parameter, we can just delete the print statements later. New code:

```py
from pedal import *

verify()

from pedal.gpt import gpt_run_prompts
gpt_run_prompts()

from pedal.resolvers import print_resolve
print_resolve()
```

**Review by acbart — COMMENTED** [review]
There are some places where the code could be decomposed a bit more, and I'm not sure the default prompts belong in a helper function, but otherwise this looks good! Good work integrating all the changes!
